[RISCV] Split regalloc between RVV and other

BeMg opened the pull request:
Enable this flow by -riscv-split-regalloc=1 (default disable), and could designate specific allocator to RVV by -riscv-rvv-regalloc=<fast|basic|greedy>

It uses the RegClass filter function to decide which regclass need to be processed.

This patch is pre-requirement for supporting PostRA vsetvl insertion pass.

BeMg commented:
This patch is for addressing the comment in https://github.com/llvm/llvm-project/pull/70549.

Seqerate the splitting register allocation part with it own patch.

BeMg commented:
Here is the SPEC2k17 benchmark result to measure the impact after implementing this patch.

with O3 and enable V extension

| Name                  | Instruction count Diff % |
| ------------------ | ---------- |
| 500.perlbench_r  | -0.03%   |
| 502.gcc_r             | -0.01%   |
| 505.mcf_r            | 0% |
| 520.omnetpp_r   | 0% |
| 523.xalancbmk_r | -0.04 % |
| 525.x264_r          | 0%|
| 531.deepsjeng_r | 0% |
| 541.leela_r          | 0% |
| 548.exchange2_r | 0% |
| 557.xz_r               | 0% |

lukel97 commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -89,6 +90,10 @@ static cl::opt<bool>
+static cl::opt<bool> EnableSplitRegAlloc("riscv-split-regalloc", cl::Hidden,
+                                   cl::desc("Enable Split RegisterAlloc for RVV"),
+                                   cl::init(false));
+
```
Should we just enable split register allocation by default? I didn't notice any regressions with this enabled on the [llvm test suite](https://github.com/llvm/llvm-project/files/13332331/regalloc-split.patch), and it doesn't seem to affect the output at all on RV64g (i.e. without v)

BeMg replied:
Thanks for reviewing.

I think the splitting register allocation only gain benefit when the following optimization is ready. Perhaps we can enable it by default when the following optimizations are stable.

preames reviewed (approved):
LGTM

topperc reviewed (approved):
LGTM

BeMg committed:
[RISCV] Split regalloc between RVV and other

Enable this flow by -riscv-split-regalloc=1 (default disable), and could designate specific allocator to RVV by
-riscv-rvv-regalloc=<fast|basic|greedy>

It uses the RegClass filter function to decide which regclass need to be processed.

This patch is pre-requirement for supporting PostRA vsetvl insertion pass.
